### ISSUE TITLE
save() Callback Returning Record Inconsistent with Database

### ISSUE THREAD

**Issue by itsalaidbacklife** [balderdashy/waterline#1091]
This issue appears similar to [975](https://github.com/balderdashy/waterline/issues/975) as well as [2073,](https://github.com/balderdashy/sails/issues/2073) but it is different in that while the save() call fails to persist the local data to the database, waterline throws no error. Essentially, when making several nested find()'s, then calling add() and remove() on different collections in the returned records and then nesting the save() calls, the local copies of the records returned by save() are **correct,** but the data is **not** updated on the database. The records returned by save() are inconsistent with the records found at localhost:1337/modelName. Is this a bug, or am I misusing save()?

Here are the relevant models:

 In Card.js:

    attributes: {
      rank: {type: 'integer'},
      suit: {type: 'integer'},
      deck: {model: 'game'},
      scrap: {model: 'game'},
      hand: {model: 'player'},
      points: {model: 'player'},
      runes: {model: 'player'},
      stackedTwo: {model: 'game'},
      targetId: {type: 'integer'},
      attached: {model: 'card'},
      attachments: {collection: 'card', via: 'attached'},

      class: {type: 'string', defaultsTo: 'card',}
    }

In Player.js:
       
    attributes: {
      socketId: {type: 'string'},
      currentGame: {model: 'game'},
      pNum: {type: 'integer'},
      hand: {collection: 'card', via: 'hand'},
      points: {collection: 'card', via: 'points'},
      runes: {collection: 'card', via: 'runes'},
      frozenId: {type: 'integer', defaultsTo: null},
      tempString: {type: 'string', defaultsTo: ''},
    }

Below is the action in our controller:
It is supposed to find two players, one of which is the 'thief' and one of which is the 'victim'. The 'theif' player removes a jack from his hand collection and adds that jack to the 'attachments' collection of a card in the 'victim' player's runes collection. The chosen card is then removed from the victim's runes and added to the 'thief' player's runes. Both players are then saved in a nested fashion, so that the callback passed to the innermost save() has access to all of the fully updated records, for the purpose of running a publishUpdate() that we do not currently call. the lines `console.log(savedP0)` and `console.log(savedP1)` print the exact expected data, but checking the database at localhost:1337/player reveals that the only change made on the database was removing the jack card from the 'thief' player's hand. None of the add() calls, nor the other remove() calls, besides the one on the 'thief' player's hand, persist on the database.

	jackBug: function (req, res) {
			Player.find([req.body.thiefId, req.body.victimId]).populateAll().exec(function (erro, players) {
				Card.findOne(req.body.targetId).populate('attachments').exec(function (err, targetCard) {
					
					if (players[0].id === req.body.thiefId) {
						var thiefIndex = 0;
						var victimIndex = 1;
					} else if (players[1].id === req.body.thiefId) {
						var thiefIndex = 1;
						var victimIndex = 0;
					}

					//Remove the jack from the theif player's hand and add the target card to the thief player's points
					players[thiefIndex].runes.add(targetCard.id);
					players[thiefIndex].hand.remove(req.body.jackId);

					//Remove the target card from the victim player's runes
					players[victimIndex].runes.remove(targetCard.id);

					//Add the jack card to the target card's attachments
					targetCard.attachments.add(req.body.jackId);

						players[0].save(function (e, savedP0) {
							players[1].save(function (e6, savedP1) {
								targetCard.save(function(e7, savedTarget) {
									console.log(savedP0);
									console.log(savedP1);
								});
							});
						});
				});
		});
	},

**Comment by particlebanana** [comment]
Are you using mongo by any chance? https://github.com/balderdashy/sails-mongo/issues/300

**Comment by itsalaidbacklife** [comment]
I am using sails-disk, not mongo. Looking at the linked issue, my problem seems different insofar as the code inside the cb passed to save actually does execute, it is only the writing to the database that appears to fail.

I am new to posting issues to github. What does the 'needs failing test' tag mean? That I should write a test using (sails-generate-test)[https://github.com/balderdashy/sails-generate-test/blob/master/FAQ.md]?

**Comment by dmarcelino** [comment]
Hi @itsalaidbacklife, "needs failing test" means this issue need a failing test PR'ed against [waterline tests](https://github.com/balderdashy/waterline/tree/master/test) or [waterline-adapter-tests](https://github.com/balderdashy/waterline-adapter-tests) (if you think it's an issue with the adapter). I don't know sails-generate-test but  it looks `sails` related, not `waterline`.

**Comment by itsalaidbacklife** [comment]
Thank you! First, I want to make sure that this is actually the place to post my issue. The project that has had this issue is actually a sails.js project. The readme for sails.js mentioned that the platform is compartmentalized into different modules and suggested posting issues related to a specific module on the corresponding github page, rather than on sails.js, itself. I posted this here because the issue appears to be with .save(), which is a part of waterline, but I am unsure whether I am having a broader issue with sails.js, itself. Should I post to the sails.js issues page, if I am unsure?

Second, I am new to testing, in general, and would greatly appreciate any resources that you could point me towards for getting sufficiently immersed to post an adequate test. Should my tests simply emulate the action that is failing, by implementing the same logic? Or should they actually lift the server, make a request, and evaluate the response? [This guide](http://sailsjs-documentation.readthedocs.org/en/latest/concepts/Testing/) to testing sails.js applications recommends using [supertest](https://github.com/visionmedia/supertest) to test controller actions by making http assertions; is that what I should be using?

**Comment by itsalaidbacklife** [comment]
Hey @dmarcelino and @particlebanana,

I have been working on getting a failing test to pull request against the waterline tests, but I am having a problem, where my controller action behaves differently during the test, than during normal use. I am using mocha, with supertest for making http requests, and bluebird for promises. 

I have refactored the jackBug action, using promises for the find() calls, and when I run the application, and use the UI to request the action, the server's local copies of the appropriate records returned in the save() callback, still have the expected values, but these changes are still not being persisted on the database, itself. However, when I use mocha to test the action, by instantiating temp data and  making a request, the data returned to save() is different than when I run the application manually. This is preventing me from writing a test that adequately captures exactly what is breaking, under normal circumstances. 

Here is the new action (with promises):

    	jackBug: function (req, res) {

		var thiefIndex;
		var victimIndex;
		var promisePlayers = new Promise(function (resolve, reject) {
			//Promise that finds both players
			Player.find({}).populateAll().exec(function (error, players) {
				if (players[0].id === req.body.thiefId) {
					thiefIndex = 0;
					victimIndex = 1;
				} else if (players[1].id === req.body.thiefId) {
					thiefIndex = 1;
					victimIndex = 0;
				}
				var sortedPlayers = [ players[thiefIndex], players[victimIndex] ];
				resolve(sortedPlayers);
			});
		});

		//Promise that finds the rune card being moved from victim.runes to thief.runes
		var promisePoints= new Promise(function (resolve, reject) {
			Card.findOne(req.body.targetId).populate('attachments').exec(function (error, rune) {
				resolve(rune);
			});
		});

		//When both promises resolve, this function executes the logic with the returned records
		Promise.all([promisePlayers, promisePoints]).then(function (values) {

			var players = values[0];
			var thief = players[0];
			var victim = players[1];
			var rune = values[1];

			thief.runes.add(rune.id);
			thief.hand.remove(req.body.jackId);
			victim.runes.remove(rune.id);
			rune.attachments.add(req.body.jackId);

			thief.save(function (error, savedThief) { //When run with the UI, savedThief is correct
				                                      //When run with mocha, savedThief is undefined
				victim.save(function (erro, savedVictim) { //savedVictim is correct when the action is run through mocha, or through the UI
					rune.save(function (err, savedRune) {  //Saved rune is correct when run with the UI, and incorrect when run with mocha
						if (thiefIndex === 0) {
							players = [thief, victim];
						} else {
							players = [victim, thief];
						}
						res.send({players: players, rune: rune});
					});
				});
			});

		});
	},

Here is what we are doing with mocha. First we call a bootstrap file (bootstrap.test.js):

    var Sails = require('sails'),
      sails;

    before(function(done) {
      this.timeout(5000);
      Sails.lift({
      }, function(err, server) {
         sails = server;

         if (err) return done(err);
         done(err, sails); 
      });
    });
    after(function(done) {
      sails.lower(done);
    });


We then run the test file: gameController.test.js (by making it the second argument of the command line):

gameController.test.js:

    var Promise = require('bluebird');
    var request = require('supertest');

    var agent = request.agent('localhost:1337'); //superTest will be used to make requests

    describe('GameController', function() {
    	describe('Emulation', function() {

		  var player0;
		  var player1;
		  var jack;
		  var rune;

		  before(function instantiateRecords(done) {
			  //Promise to create a temp player for testing
			  var promisePlayer0 = new Promise(function(resolve, reject) {
				  Player.create({pNum: 0}, function(error, createdPlayer0) {
					  return resolve(createdPlayer0);
				  });
			  });

			  promisePlayer0.then(function (createdPlayer0) {
				  player0 = createdPlayer0;
			  });

			  //Promise to create a second temp player
			  var promisePlayer1 = new Promise(function (resolve, reject) {
				  Player.create({pNum: 1}, function (error, createdPlayer1) {
					  return resolve(createdPlayer1);
				  });
			  });

			  promisePlayer1.then(function (createdPlayer1) {
				  player1 = createdPlayer1;
			  });

			  //Promise to create a jack card to be added to player0's hand collection
			  var promiseJack = new Promise(function (resolve, reject) {
				  Card.create({suit: 3, rank: 11}, function (error, createdJack) {
					  return resolve(createdJack);
				  });
			  });

			  promiseJack.then(function (createdJack) {
				  jack = createdJack;
			  });

			  //Promise to create a King card to be added to player1's runes collection
			  var promiseRune = new Promise(function (resolve, reject) {
				  Card.create({suit: 3, rank: 13}, function (error, createdRune) {
					  return resolve(createdRune);
				  });
			  });

			  promiseRune.then(function (createdRune) {
				  rune = createdRune;
			  });

			  //When all players and cards are created, add the cards to the appropriate hand and runes collections of the players
			  Promise.all([promisePlayer0, promisePlayer1, promiseJack, promiseRune]).then(function (values) {

				  player0.hand.add(jack.id);
				  player1.runes.add(rune.id);

				  player0.save(function (error, savedPlayer0) {
					  player1.save(function (erro, savedPlayer1) {
						  player0 = savedPlayer0;
						  player1 = savedPlayer1
						  done();
					  });
				  });
				
			  });

		  });
              //After all record initialization, request the jackBug action
		  it('should jack a rune', function(done) {
			  agent.post('/game/jackBug').send({thiefId: player0.id, victimId: player1.id, targetId: rune.id}).end(function (error, res) {
				  console.log(res.body);
				  done();
			  });
		  });

	  });
    });

When the test runs, the jackBug action behaves differently. The records are still found properly, at the beginning, but inside of the save() callbacks, the first record to be saved (in this case, savedThief), is undefined. Am I approaching the testing wrong, fundamentally? Should I be making a request, having the server send back a response with various records and apply logic to determine if the test failed in gameController.test.js?

Is it fundamentally wrong, to nest the save() calls like this, in order to gain access to all updated records in the scope of one function? If so, how am I supposed to do that? I have tried using a different promise for each save() call, and using a Promise.all() to return the updated records into one function, but this has also not worked. When I did this, the records returned to the Promise.all() were incorrect in the same way that the database is always incorrect. This made the local server data consistent with the database, but it still didn't properly update the values.

**Comment by itsalaidbacklife** [comment]
Okay, I have figured it out! There are two issues that were occurring: First and foremost, I had a confusion about how .add(), .remove(), and .save() work in tandem. The fundamental problem is that each player's runes collection was  a **one-to-many** association with the various card records, and I was calling .add() and .remove() on two different players' runes collections and moving the same card with both calls. This doesn't work because both of those collections are associated via the **same foreign key** on the card record. What I didn't understand is that what .remove() actually **does** is cue up a cb (that is fired when .save() is called) that goes to the associated (the card) record and **deletes the foreign key to the record (the player) in whose collection the removed record (the card) resides**. This means that my .add() call was being overwritten by my .remove() call, because .add() was instantiating a foreign key from the card to the player, and then .remove() was deleting that foreign key, **even though it no longer pointed to the player from whose runes collection I was attempting to remove the card**. The fix is very simple: simply don't call .remove() at all, as using .add() to push the card into one player's runes collection automatically removes it from the other player's runes by overwriting the foreign key. 

There was a second, and indirect problem that this brought to my attention, however, which is that a record returned inside a .save() callback is **not always consistent with the changes made to the database**. I believe that the issue was that .save() was returning a populated player record that matched what we would expect the database to look like, based on the .add() and .remove() calls, but it was failing to take into account the faulty logic that I was using, and it therefore returned a record that looked right, even though the database was wrong. Is this what was happening, and should .save() be tweaked to always return exactly what the database has stored?

I have been able to make the record returned by .save() match the database (even when the .add and remove calls overwrite each other), by leveraging promises with Bluebird.js. There is probably a better way to do it, but I make a new promise for each .find(), return the found record when the promise fulfills, use a Promise.all() to wait for all records to be found before continuing with the logic, then use a .spread() to perform all the saves, and wait for them to return before calling the publishUpdate. It seems that noone else has had my foolish problem, so this pattern may not prove helpful to anyone, else, but I'm closing the issue now that it seems to be all figured out!
